schema/crdb/dbinit.sql

```diff
@@ -4063,6 +4063,20 @@ VALUES (
 ON CONFLICT (id)
 DO NOTHING;
 
+CREATE TYPE IF NOT EXISTS omicron.public.migration_state AS ENUM (
+  'in_progress',
+  'failed',
+  'completed'
+);
+
+-- A table of the states of current migrations.
+CREATE TABLE IF NOT EXISTS omicron.public.migration (
+    id UUID PRIMARY KEY,
+    source_state omicron.public.migration_state NOT NULL,
+    source_propolis_id UUID NOT NULL,
+    target_state omicron.public.migration_state NOT NULL,
+    target_propolis_id UUID NOT NULL
```

Comment:
We mostly covered this offline, but for posterity: I think we'll also want generation numbers and update times, and they need to be recorded separately for the source and target.

My mental model is that migration state is adjacent to VMM state and will be pushed to (or pulled by) Nexus alongside VMM state updates. That is:

1. on a migration phase change, Propolis emits new migration state to any callers of its instance watcher endpoint
2. sled agent observes these changes and reflects them up to Nexus via the same "vmm state changed" endpoint it uses for "major" state changes
3. in response, Nexus writes the new vmm/migration states to CRDB
4. Nexus decides whether the change it recorded was momentous enough to require an instance update saga

It's only safe to do this if we can guarantee that Nexus receives and processes migration state changes in order. I don't think that's the case today because of the instance watcher task, but even if that task weren't there, I think it's better to give sled agent and Nexus some flexibility in how these messages are sent and processed (e.g. we might someday want to make the "write back to the DB" part of the sled agent upcall asynchronous, which would allow sled agent to come back, read some even newer state from Propolis, and send that up to Nexus before the prior state was saved).

The source- and target-reported states are managed by independent sleds that won't be able to see each other's generation numbers, so they should be separate in the DB as well (as should their update times).

We'll have a little bit of work to do to merge states and update times if/when we expose migration status via API, but I don't think it will be anything intractable.

Reply:
Yup, I agree; working on that now. I'm not totally convinced we need both update times _and_ generation counts if the generation is owned uniquely by the sled-agent, but I suppose we may as well add the timestamps as well.

Reply:
If we don't have separate timestamps, I think things can get out of whack if there's a delay between the moment the timestamp is captured for some record we're about to write and the moment Nexus writes it back:

t=1: Nexus captures time for a write that says "source = in_progress"
t=2: Nexus captures time for a write that says "target = done"
t=3: Nexus writes "source = in_progress, updated_time = 1"
<time passes>
t=100: Nexus prepares to write "source = done"
t=101: Nexus writes "source = done, updated_time = 100"
t=102: Nexus writes "target = done, updated_time = 2"

I'm not sure if using CRDB's `now()` function as a time source would help with this. There's still the possibility of a delay between the time CRDB gets the timestamp and the time it commits it, but it's possible that the conflicting writes would cause a transaction retry (giving CRDB a chance to refresh the timestamp).

Probably simpler, I think, just to have two timestamps (with the added advantage that we'll be able to see exactly how long one or the other side of a migration appears to have been out to lunch).

Reply:
I totally agree that we may as well record timestamps, especially for diagnostic purposes, and that if we're going to have them, we need to have two of them. What I meant was that, AFAIK, the generation numbers are sufficient to ensure *correctness* here (such as in the case you described). Am I overlooking something?

Reply:
Oh, my mistake, I misread you. Yeah, the generation numbers are really all you need for this, at least insofar as "older states shouldn't overtake newer ones" is concerned.